tests/Extensions/JobRetryExtensionTest.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace SilverStripe\Forager\Tests\Extensions;
+
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Forager\Tests\Fake\BatchJobFake;
+use SilverStripe\ORM\FieldType\DBDatetime;
+use Symbiote\QueuedJobs\DataObjects\QueuedJobDescriptor;
+use Symbiote\QueuedJobs\Services\QueuedJob;
+use Symbiote\QueuedJobs\Services\QueuedJobService;
+
+class JobRetryExtensionTest extends SapphireTest
+{
+
+    protected $usesDatabase = true; // phpcs:ignore SlevomatCodingStandard.TypeHints
+
+    /**
+     * We'll run the job a total of 4 times. We expect 3 attempts to break, but be retried, and we expect the 4th try
+     * to break and remain broken
+     *
+     * @dataProvider provideCodes
+     */
+    public function testRetryCodes(int $code): void
```

Comment:
does this test need `Config::modify()->set(QueuedJobService::class, 'use_shutdown_function', false);`?

Reply:
Ah yea, that old chestnut. I haven't seen any issues yet, but I'll add it now anyway.